x-pack/plugins/apm/server/lib/services/get_service_details.ts

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import {
+  AGENT,
+  CLOUD,
+  CLOUD_AVAILABILITY_ZONE,
+  CLOUD_MACHINE_TYPE,
+  CONTAINER,
+  HOST,
+  KUBERNETES,
+  PROCESSOR_EVENT,
+  SERVICE,
+  SERVICE_NAME,
+  SERVICE_NODE_NAME,
+  SERVICE_VERSION,
+} from '../../../common/elasticsearch_fieldnames';
+import { ProcessorEvent } from '../../../common/processor_event';
+import { rangeFilter } from '../../../common/utils/range_filter';
+import { TransactionRaw } from '../../../typings/es_schemas/raw/transaction_raw';
+import { getBucketSize } from '../helpers/get_bucket_size';
+import { Setup, SetupTimeRange } from '../helpers/setup_request';
+
+type ServiceDetails = Pick<
+  TransactionRaw,
+  'service' | 'agent' | 'host' | 'container' | 'kubernetes' | 'cloud'
+>;
+
+interface ServiceDetailsResponse {
+  service?: {
+    versions?: string[];
+    runtime?: {
+      name: string;
+      version: string;
+    };
+    framework?: string;
+    agent: {
+      name: string;
+      version: string;
+    };
+  };
+  container?: {
+    os?: string;
+    isContainerized?: boolean;
+    avgNumberInstances?: number;
+    orchestration?: 'Kubernetes' | 'Docker';
+  };
+  cloud?: {
+    provider?: string;
+    availabilityZones?: string[];
+    machineTypes?: string[];
+    projectName?: string;
+  };
+}
+
+export async function getServiceDetails({
+  serviceName,
+  setup,
+}: {
+  serviceName: string;
+  setup: Setup & SetupTimeRange;
+}): Promise<ServiceDetailsResponse> {
+  const { start, end, apmEventClient } = setup;
+  const { intervalString } = getBucketSize({ start, end });
+
+  const filter = [
+    { term: { [SERVICE_NAME]: serviceName } },
+    { term: { [PROCESSOR_EVENT]: ProcessorEvent.transaction } },
+    { range: rangeFilter(start, end) },
+    ...setup.esFilter,
+  ];
+
+  const should = [
+    CONTAINER,
+    KUBERNETES,
+    CLOUD,
+    HOST,
+    AGENT,
+  ].map((fieldName) => ({ exists: { field: fieldName } }));
```

Comment:
nit: I'm not convinced that the byte savings are worth making this dynamic - static is (almost) always better than dynamic.

```suggestion
const should = [
  { exists: { field: CONTAINER } },
  { exists: { field: KUBERNETES } },
  { exists: { field: CLOUD } },
  { exists: { field: HOST } },
  { exists: { field: AGENT } },
];
```

But I think it's a coin toss so if you prefer the other approach that's fine.